# ML4SCI QMLHEP GSoC 2026 — Quantum Circuit Design with LLMs
## Evaluation Tasks: Agentic Framework with Orchestral AI + Gemini

**Author:** Leo Chang  
**Date:** 2026-03-28  
**LLM Backend:** Google Gemini 2.5 Flash  
**Agent Framework:** Orchestral AI 1.3.0  
**Quantum Backend:** PennyLane 0.44.1  

### Summary

This notebook implements an agentic framework for quantum circuit design using Orchestral AI with Google Gemini as the LLM backend. Three tasks are demonstrated:

**Task 1 — Hello World Tools:** Four quantum-related tools (`hilbert_dim`, `describe_quantum_gate`, `generate_circuit_code`, `estimate_circuit_resources`) are defined and tested through 5 agent queries, including a multi-call query.

**Task 2 — QNN Training Tool:** A hybrid quantum neural network (classical encoder + 4-qubit VQC + classical head) for MNIST classification is wrapped as an agent tool. The agent trains the model with different hyperparameters and analyzes learning dynamics.

**Task 3 — Agent-Driven HPO:** The agent autonomously optimizes the learning rate over 6 trials, demonstrating explore-then-exploit behavior. It identifies lr=0.0075 as optimal (test_acc=40.5%), providing reasoning before each trial and a comprehensive final report.

---
## Setup

In [ ]:
import os
import sys
import json
import time
import numpy as np

# ── API Key ──
os.environ["GOOGLE_API_KEY"] = "AIzaSyBp55xcQmTExVJSP7uBxg0djNQdXCpE0m0"

# ── Imports ──
from orchestral import Agent, define_tool
from orchestral.llm import Gemini
import pennylane as qml
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset

print(f"Python       : {sys.version}")
print(f"PennyLane    : {qml.__version__}")
print(f"PyTorch      : {torch.__version__}")
print(f"Orchestral AI: OK")
print(f"LLM Backend  : Google Gemini")
print(f"CUDA         : {torch.cuda.is_available()}")

Python       : 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0]
PennyLane    : 0.44.1
PyTorch      : 2.11.0+cu130
Orchestral AI: OK
LLM Backend  : Google Gemini
CUDA         : True


---
# Task 1: Orchestral AI Setup & Hello World Tools

Four quantum-related tools are defined, each with clear docstrings. The agent is tested with 5 queries targeting different tools.

### Tool Definitions

In [ ]:
@define_tool()
def hilbert_dim(n_qubits: int) -> dict:
    """Compute the Hilbert space dimension for a given number of qubits.

    Given n_qubits, this tool returns:
    - hilbert_dimension: the size of the Hilbert space (2^n)
    - ram_bytes: memory needed to store the full statevector (16 bytes per complex amplitude)
    - ram_readable: human-readable RAM string (KB/MB/GB/TB)
    - classically_simulable: whether the system can be simulated on a classical computer (n <= 28)

    Use this tool when you need to assess the feasibility of simulating a quantum system.

    Args:
        n_qubits: Number of qubits in the quantum system (positive integer)

    Returns:
        Dictionary with hilbert_dimension, ram_bytes, ram_readable, classically_simulable
    """
    dim = 2 ** n_qubits
    ram = dim * 16  # 16 bytes per complex128 amplitude

    if ram < 1024:
        ram_str = f"{ram} B"
    elif ram < 1024**2:
        ram_str = f"{ram / 1024:.1f} KB"
    elif ram < 1024**3:
        ram_str = f"{ram / 1024**2:.1f} MB"
    elif ram < 1024**4:
        ram_str = f"{ram / 1024**3:.1f} GB"
    else:
        ram_str = f"{ram / 1024**4:.1f} TB"

    return {
        "n_qubits": n_qubits,
        "hilbert_dimension": dim,
        "ram_bytes": ram,
        "ram_readable": ram_str,
        "classically_simulable": n_qubits <= 28,
    }

print("Tool 1 [hilbert_dim] defined")

Tool 1 [hilbert_dim] defined


In [ ]:
GATE_DATABASE = {
    "H": {
        "name": "Hadamard",
        "n_qubits": 1,
        "parametric": False,
        "matrix": "1/sqrt(2) * [[1, 1], [1, -1]]",
        "description": "Creates equal superposition. Maps |0> to |+> and |1> to |->.",
    },
    "X": {
        "name": "Pauli-X (NOT)",
        "n_qubits": 1,
        "parametric": False,
        "matrix": "[[0, 1], [1, 0]]",
        "description": "Bit flip gate. Maps |0> to |1> and |1> to |0>.",
    },
    "Y": {
        "name": "Pauli-Y",
        "n_qubits": 1,
        "parametric": False,
        "matrix": "[[0, -i], [i, 0]]",
        "description": "Combines bit flip and phase flip.",
    },
    "Z": {
        "name": "Pauli-Z",
        "n_qubits": 1,
        "parametric": False,
        "matrix": "[[1, 0], [0, -1]]",
        "description": "Phase flip gate. Maps |1> to -|1>.",
    },
    "CNOT": {
        "name": "Controlled-NOT (CX)",
        "n_qubits": 2,
        "parametric": False,
        "matrix": "[[1,0,0,0],[0,1,0,0],[0,0,0,1],[0,0,1,0]]",
        "description": "Flips target qubit if control qubit is |1>. Primary entangling gate.",
    },
    "CZ": {
        "name": "Controlled-Z",
        "n_qubits": 2,
        "parametric": False,
        "matrix": "[[1,0,0,0],[0,1,0,0],[0,0,1,0],[0,0,0,-1]]",
        "description": "Applies Z to target if control is |1>. Symmetric entangling gate.",
    },
    "RX": {
        "name": "Rotation-X",
        "n_qubits": 1,
        "parametric": True,
        "matrix": "[[cos(t/2), -i*sin(t/2)], [-i*sin(t/2), cos(t/2)]]",
        "description": "Parametric rotation around X-axis by angle theta. Used in VQCs.",
    },
    "RY": {
        "name": "Rotation-Y",
        "n_qubits": 1,
        "parametric": True,
        "matrix": "[[cos(t/2), -sin(t/2)], [sin(t/2), cos(t/2)]]",
        "description": "Parametric rotation around Y-axis by angle theta. Most common VQC gate.",
    },
    "RZ": {
        "name": "Rotation-Z",
        "n_qubits": 1,
        "parametric": True,
        "matrix": "[[exp(-it/2), 0], [0, exp(it/2)]]",
        "description": "Parametric rotation around Z-axis by angle theta. Used in VQCs.",
    },
    "CRZ": {
        "name": "Controlled-RZ",
        "n_qubits": 2,
        "parametric": True,
        "matrix": "Controlled version of RZ gate",
        "description": "Applies RZ(theta) to target if control is |1>. Soft entangling gate.",
    },
    "SWAP": {
        "name": "SWAP",
        "n_qubits": 2,
        "parametric": False,
        "matrix": "[[1,0,0,0],[0,0,1,0],[0,1,0,0],[0,0,0,1]]",
        "description": "Swaps the states of two qubits.",
    },
    "T": {
        "name": "T (pi/8)",
        "n_qubits": 1,
        "parametric": False,
        "matrix": "[[1, 0], [0, exp(i*pi/4)]]",
        "description": "Phase gate with pi/4 phase. Important for universal gate sets.",
    },
    "S": {
        "name": "S (Phase)",
        "n_qubits": 1,
        "parametric": False,
        "matrix": "[[1, 0], [0, i]]",
        "description": "Phase gate with pi/2 phase. S^2 = Z.",
    },
    "TOFFOLI": {
        "name": "Toffoli (CCX)",
        "n_qubits": 3,
        "parametric": False,
        "matrix": "8x8 identity with bottom-right 2x2 block = [[0,1],[1,0]]",
        "description": "Flips target if both controls are |1>. Universal for classical computation.",
    },
}


@define_tool()
def describe_quantum_gate(gate_name: str) -> dict:
    """Look up information about a quantum gate by name.

    Returns the gate's full name, matrix representation, number of qubits it acts on,
    whether it is parametric (has tunable parameters), and a description of its action.

    Supported gates: H, X, Y, Z, CNOT, CZ, RX, RY, RZ, CRZ, SWAP, T, S, TOFFOLI.

    Args:
        gate_name: Name of the quantum gate (case-insensitive). Examples: "CNOT", "RY", "H"

    Returns:
        Dictionary with name, n_qubits, parametric, matrix, description
    """
    key = gate_name.strip().upper()
    if key in GATE_DATABASE:
        info = GATE_DATABASE[key]
        return {
            "gate": key,
            "full_name": info["name"],
            "n_qubits": info["n_qubits"],
            "parametric": info["parametric"],
            "matrix": info["matrix"],
            "description": info["description"],
        }
    else:
        available = sorted(GATE_DATABASE.keys())
        return {
            "error": f"Gate '{gate_name}' not found.",
            "available_gates": available,
        }

print("Tool 2 [describe_quantum_gate] defined")

Tool 2 [describe_quantum_gate] defined


In [ ]:
@define_tool()
def generate_circuit_code(
    n_qubits: int,
    n_layers: int,
    rotation_gate: str,
    entangler_gate: str,
) -> dict:
    """Generate PennyLane Python code for a variational quantum circuit (VQC).

    Creates a parameterized quantum circuit with the specified architecture:
    - Each layer applies AngleEmbedding for data encoding, then rotation gates
      on every qubit, then entangling gates in a linear chain.
    - The circuit measures expectation values of PauliZ on all qubits.

    Args:
        n_qubits: Number of qubits (recommended: 2-8)
        n_layers: Number of variational layers (recommended: 1-6)
        rotation_gate: Single-qubit rotation gate to use. Options: "RY", "RX", "RZ"
        entangler_gate: Two-qubit entangling gate. Options: "CNOT", "CZ", "CRZ", "none"

    Returns:
        Dictionary with:
        - code: Complete PennyLane Python code as a string
        - n_params: Total number of trainable parameters
        - architecture_summary: Human-readable description of the circuit
    """
    rot = rotation_gate.strip().upper()
    ent = entangler_gate.strip().upper()

    if rot not in ("RY", "RX", "RZ"):
        return {"error": f"Invalid rotation_gate '{rotation_gate}'. Use RY, RX, or RZ."}
    if ent not in ("CNOT", "CZ", "CRZ", "NONE"):
        return {"error": f"Invalid entangler_gate '{entangler_gate}'. Use CNOT, CZ, CRZ, or none."}

    n_params = n_layers * n_qubits
    if ent == "CRZ":
        n_params += n_layers * max(0, n_qubits - 1)

    # Build code string
    lines = []
    lines.append("import pennylane as qml")
    lines.append("import numpy as np")
    lines.append("")
    lines.append(f"n_qubits = {n_qubits}")
    lines.append(f"dev = qml.device('default.qubit', wires=n_qubits)")
    lines.append("")
    lines.append("@qml.qnode(dev)")
    lines.append("def circuit(inputs, weights):")
    lines.append(f'    """VQC: {n_layers} layer(s), {rot} rotation, {ent} entangler, {n_qubits} qubits"""')

    lines.append(f"    for layer in range({n_layers}):")
    lines.append(f"        qml.AngleEmbedding(inputs, wires=range({n_qubits}))")

    # Rotation gates
    lines.append(f"        for q in range({n_qubits}):")
    lines.append(f"            qml.{rot}(weights[layer, q], wires=q)")

    # Entangling gates
    if ent == "NONE":
        lines.append("        # No entanglement")
    elif ent in ("CNOT", "CZ"):
        lines.append(f"        for q in range({n_qubits} - 1):")
        lines.append(f"            qml.{ent}(wires=[q, q + 1])")
    elif ent == "CRZ":
        lines.append(f"        for q in range({n_qubits} - 1):")
        lines.append(f"            qml.CRZ(weights[layer, {n_qubits} + q], wires=[q, q + 1])")

    lines.append("")
    lines.append(f"    return [qml.expval(qml.PauliZ(q)) for q in range({n_qubits})]")
    lines.append("")
    lines.append(f"# Total trainable parameters: {n_params}")
    lines.append(f"# Weight shape: ({n_layers}, {n_qubits if ent != 'CRZ' else n_qubits + max(0, n_qubits - 1)})")

    code = "\n".join(lines)

    summary = (
        f"{n_qubits}-qubit VQC with {n_layers} layer(s), "
        f"{rot} rotation gates, {ent} entangler, "
        f"{n_params} trainable parameters"
    )

    return {
        "code": code,
        "n_params": n_params,
        "n_qubits": n_qubits,
        "n_layers": n_layers,
        "rotation_gate": rot,
        "entangler_gate": ent,
        "architecture_summary": summary,
    }

print("Tool 3 [generate_circuit_code] defined")

Tool 3 [generate_circuit_code] defined


In [ ]:
@define_tool()
def estimate_circuit_resources(
    n_qubits: int,
    n_layers: int,
    entangler_gate: str,
) -> dict:
    """Estimate the computational resources required to run a variational quantum circuit.

    Computes gate counts, circuit depth, number of two-qubit gates, and estimates
    simulation time on a classical computer. Useful for deciding circuit complexity
    before building.

    Args:
        n_qubits: Number of qubits (positive integer)
        n_layers: Number of variational layers (positive integer)
        entangler_gate: Two-qubit entangling gate. Options: "CNOT", "CZ", "CRZ", "none"

    Returns:
        Dictionary with:
        - total_gates: Total number of quantum gates
        - single_qubit_gates: Number of single-qubit gates
        - two_qubit_gates: Number of two-qubit gates
        - circuit_depth: Estimated circuit depth (critical layers in dependency graph)
        - n_params: Number of trainable parameters
        - estimated_sim_time: Rough simulation time estimate
        - nisq_feasibility: Whether this circuit is practical on current NISQ hardware
    """
    ent = entangler_gate.strip().upper()
    if ent not in ("CNOT", "CZ", "CRZ", "NONE"):
        return {"error": f"Invalid entangler_gate '{entangler_gate}'. Use CNOT, CZ, CRZ, or none."}

    # Per layer: n_qubits angle embeddings + n_qubits rotations
    single_q_per_layer = 2 * n_qubits  # embedding + rotation
    two_q_per_layer = (n_qubits - 1) if ent != "NONE" else 0

    total_single = single_q_per_layer * n_layers
    total_two = two_q_per_layer * n_layers
    total_gates = total_single + total_two

    # Depth: each layer has ~3 depth levels (embedding, rotation, entangling chain)
    depth_per_layer = 2 + (n_qubits - 1 if ent != "NONE" else 0)
    total_depth = depth_per_layer * n_layers

    # Parameters
    n_params = n_layers * n_qubits
    if ent == "CRZ":
        n_params += n_layers * (n_qubits - 1)

    # Simulation time estimate (rough, based on statevector)
    dim = 2 ** n_qubits
    ops = total_gates * dim * dim  # matrix-vector multiplications
    flops_per_sec = 1e9  # ~1 GFLOP for Python/numpy
    est_time = ops / flops_per_sec

    if est_time < 1:
        time_str = f"{est_time * 1000:.1f} ms"
    elif est_time < 60:
        time_str = f"{est_time:.1f} s"
    elif est_time < 3600:
        time_str = f"{est_time / 60:.1f} min"
    else:
        time_str = f"{est_time / 3600:.1f} hours"

    # NISQ feasibility: depth < 100, 2q gates < 1000, qubits < 50
    nisq_ok = total_depth < 100 and total_two < 1000 and n_qubits < 50

    return {
        "n_qubits": n_qubits,
        "n_layers": n_layers,
        "entangler_gate": ent,
        "total_gates": total_gates,
        "single_qubit_gates": total_single,
        "two_qubit_gates": total_two,
        "circuit_depth": total_depth,
        "n_params": n_params,
        "estimated_sim_time": time_str,
        "nisq_feasibility": nisq_ok,
        "nisq_notes": (
            "Feasible on current NISQ devices"
            if nisq_ok
            else "Too deep or too many 2-qubit gates for reliable NISQ execution"
        ),
    }

print("Tool 4 [estimate_circuit_resources] defined")

Tool 4 [estimate_circuit_resources] defined


## Direct Tool Verification (No API Calls)

Before testing with the agent, we verify each tool works correctly by calling them directly.

In [ ]:
print("=" * 60)
print("DIRECT TOOL VERIFICATION (no API calls)")
print("=" * 60)

print("\n--- Tool 1: hilbert_dim ---")
r1 = hilbert_dim.execute(n_qubits=4)
print(r1)

print("\n--- Tool 2: describe_quantum_gate ---")
r2 = describe_quantum_gate.execute(gate_name="CNOT")
print(r2)

print("\n--- Tool 3: generate_circuit_code ---")
r3 = generate_circuit_code.execute(
    n_qubits=4, n_layers=2, rotation_gate="RY", entangler_gate="CNOT"
)
print(r3)

print("\n--- Tool 4: estimate_circuit_resources ---")
r4 = estimate_circuit_resources.execute(
    n_qubits=4, n_layers=2, entangler_gate="CNOT"
)
print(r4)

print("\nAll 4 tools verified successfully.")

--- Tool 1: hilbert_dim ---
{'n_qubits': 4, 'hilbert_dimension': 16, 'ram_bytes': 256, 'ram_readable': '256 B', 'classically_simulable': True}

--- Tool 2: describe_quantum_gate ---
{'gate': 'CNOT', 'full_name': 'Controlled-NOT (CX)', 'n_qubits': 2, 'parametric': False, 'matrix': '[[1,0,0,0],[0,1,0,0],[0,0,0,1],[0,0,1,0]]', 'description': 'Flips target qubit if control qubit is |1>. Primary entangling gate.'}

--- Tool 3: generate_circuit_code ---
{'code': 'import pennylane as qml\nimport numpy as np\n\nn_qubits = 4\ndev = qml.device(\'default.qubit\', wires=n_qubits)\n\n@qml.qnode(dev)\ndef circuit(inputs, weights):\n    """VQC: 2 layer(s), RY rotation, CNOT entangler, 4 qubits"""\n    for layer in range(2):\n        qml.AngleEmbedding(inputs, wires=range(4))\n        for q in range(4):\n            qml.RY(weights[layer, q], wires=q)\n        for q in range(4 - 1):\n            qml.CNOT(wires=[q, q + 1])\n\n    return [qml.expval(qml.PauliZ(q)) for q in range(4)]\n\n# Total trainable 

## Agent Demo — Gemini 2.5 Flash + 4 Tools

We create an agent powered by Google Gemini 2.5 Flash and test it with 5 natural language queries. Each query targets a different tool, and the final query tests multi-tool calling (agent must invoke `hilbert_dim` three separate times).

In [ ]:
ALL_TOOLS = [hilbert_dim, describe_quantum_gate, generate_circuit_code, estimate_circuit_resources]

agent = Agent(
    llm=Gemini(model="gemini-2.5-flash"),
    tools=ALL_TOOLS,
)

queries = [
    "How large is the Hilbert space for a 4-qubit system? Can it be simulated on a classical computer?",
    "What is the CNOT gate? Show me its matrix and explain how many qubits it acts on.",
    "Generate PennyLane code for a variational quantum circuit with 4 qubits, 2 layers, using RY rotation gates and CNOT entanglers.",
    "If I build a circuit with 6 qubits, 3 layers, and CZ entanglers, how many gates will it have and is it feasible on NISQ hardware?",
    "Compare the simulation feasibility for 10, 20, and 50 qubit systems. For each, tell me the Hilbert space size and whether it can be classically simulated.",
]

for i, q in enumerate(queries, 1):
    print(f"\n{chr(9472) * 60}")
    print(f"Query {i}: {q}")
    print(f"{chr(9472) * 60}")
    response = agent.run(q, max_iterations=10)
    print(f"\nAgent Response:\n{response.text}")
    print(f"\n[Cost so far: ${agent.get_total_cost():.4f}]")

print(f"\n{'=' * 60}")
print(f"TOTAL API COST: ${agent.get_total_cost():.4f}")
print(f"{'=' * 60}")

────────────────────────────────────────────────────────────
Query 1: How large is the Hilbert space for a 4-qubit system? Can it be simulated on a classical computer?
────────────────────────────────────────────────────────────

Agent Response:
A 4-qubit system has a Hilbert space dimension of 16. Yes, it can be simulated on a classical computer.

[Cost so far: $0.0004]

────────────────────────────────────────────────────────────
Query 2: What is the CNOT gate? Show me its matrix and explain how many qubits it acts on.
────────────────────────────────────────────────────────────

Agent Response:
The CNOT gate, also known as Controlled-NOT (CX), is a primary entangling gate. It acts on 2 qubits. Its matrix representation is:

[[1, 0, 0, 0],
 [0, 1, 0, 0],
 [0, 0, 0, 1],
 [0, 0, 1, 0]]

It flips the target qubit if the control qubit is in the |1> state.

[Cost so far: $0.0008]

────────────────────────────────────────────────────────────
Query 3: Generate PennyLane code for a variation

## Results Summary

| Query | Target Tool | Agent Called Correctly | Key Result |
|-------|-------------|----------------------|------------|
| 1. Hilbert space for 4 qubits | `hilbert_dim` | Yes | dim=16, simulable=True |
| 2. CNOT gate info | `describe_quantum_gate` | Yes | 2-qubit, matrix shown |
| 3. Generate VQC code | `generate_circuit_code` | Yes | 8 params, full code |
| 4. Resource estimation | `estimate_circuit_resources` | Yes | 51 gates, NISQ feasible |
| 5. Compare 10/20/50 qubits | `hilbert_dim` x 3 | Yes | Multi-call success |

**Total API cost:** $0.0032

All 4 tools were reliably invoked by the agent. Query 5 demonstrates the agent's ability to autonomously make multiple tool calls to answer a comparative question.

---
# Shared Infrastructure for Task 2 & 3

The following dataset loader and QNN architecture are shared between Task 2 and Task 3.

## Dataset — MNIST Subset

800 training / 200 test images, 10 classes.

In [ ]:
def get_loaders(n_train=800, n_test=200, batch_size=32):
    """Load MNIST subset. Falls back to synthetic data if MNIST unavailable."""
    try:
        from torchvision import datasets, transforms
        transform = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize((0.1307,), (0.3081,))
        ])
        train_ds = datasets.MNIST("./data", train=True, download=True, transform=transform)
        test_ds = datasets.MNIST("./data", train=False, download=True, transform=transform)

        # Subset
        torch.manual_seed(42)
        train_idx = torch.randperm(len(train_ds))[:n_train]
        test_idx = torch.randperm(len(test_ds))[:n_test]

        train_x = torch.stack([train_ds[i][0] for i in train_idx])
        train_y = torch.tensor([train_ds[i][1] for i in train_idx])
        test_x = torch.stack([test_ds[i][0] for i in test_idx])
        test_y = torch.tensor([test_ds[i][1] for i in test_idx])
        print(f"MNIST loaded: {n_train} train, {n_test} test")

    except Exception as e:
        print(f"MNIST unavailable ({e}), using synthetic data")
        torch.manual_seed(42)
        train_x = torch.randn(n_train, 1, 28, 28)
        train_y = torch.randint(0, 10, (n_train,))
        test_x = torch.randn(n_test, 1, 28, 28)
        test_y = torch.randint(0, 10, (n_test,))

    train_loader = DataLoader(
        TensorDataset(train_x.view(n_train, -1), train_y),
        batch_size=batch_size, shuffle=True
    )
    test_loader = DataLoader(
        TensorDataset(test_x.view(n_test, -1), test_y),
        batch_size=batch_size, shuffle=False
    )
    return train_loader, test_loader

TRAIN_LOADER, TEST_LOADER = get_loaders()

MNIST loaded: 800 train, 200 test


## QNN Architecture — Fixed Hybrid Model

| Layer | Details |
|-------|--------|
| Classical Encoder | 784 -> 64 (ReLU) -> 4 (Tanh) |
| Quantum Layer | 4-qubit VQC, 2 layers, RY rotation + CNOT entanglement |
| Classical Head | 4 -> 10 (logits) |

In [ ]:
N_QUBITS = 4

dev = qml.device("default.qubit", wires=N_QUBITS)

@qml.qnode(dev, interface="torch", diff_method="backprop")
def quantum_circuit(inputs, weights):
    """4-qubit VQC: 2 layers, RY rotation, CNOT entanglement."""
    n_layers = 2
    for layer in range(n_layers):
        qml.AngleEmbedding(inputs, wires=range(N_QUBITS))
        for q in range(N_QUBITS):
            qml.RY(weights[layer, q], wires=q)
        for q in range(N_QUBITS - 1):
            qml.CNOT(wires=[q, q + 1])
    return [qml.expval(qml.PauliZ(i)) for i in range(N_QUBITS)]


class HybridQNN(nn.Module):
    """Hybrid quantum-classical neural network for 10-class classification.

    Architecture:
        Classical Encoder: 784 -> 64 (ReLU) -> 4 (Tanh)
        Quantum Layer:     4-qubit VQC, 2 layers, RY+CNOT, 8 parameters
        Classical Head:    4 -> 10 (logits)
    """
    def __init__(self):
        super().__init__()
        self.encoder = nn.Sequential(
            nn.Linear(784, 64),
            nn.ReLU(),
            nn.Linear(64, N_QUBITS),
            nn.Tanh(),
        )
        weight_shapes = {"weights": (2, N_QUBITS)}
        self.qlayer = qml.qnn.TorchLayer(quantum_circuit, weight_shapes)
        self.head = nn.Linear(N_QUBITS, 10)

    def forward(self, x):
        x = self.encoder(x)
        x = self.qlayer(x)
        x = self.head(x)
        return x

print(f"QNN Architecture:")
print(f"  Encoder:  784 -> 64 -> {N_QUBITS}")
print(f"  Quantum:  {N_QUBITS}-qubit VQC, 2 layers, RY+CNOT")
print(f"  Head:     {N_QUBITS} -> 10")
_tmp = HybridQNN()
_n_params = sum(p.numel() for p in _tmp.parameters())
print(f"  Total parameters: {_n_params}")
del _tmp

QNN Architecture:
  Encoder:  784 -> 64 -> 4
  Quantum:  4-qubit VQC, 2 layers, RY+CNOT
  Head:     4 -> 10
  Total parameters: 50558


---
# Task 2: Tool for Training a Hybrid QNN

The training logic is wrapped as an Orchestral AI tool (`train_qnn`). The agent can specify `epochs` and `learning_rate` and receives structured metrics including per-epoch history for learning curve analysis.

## Tool Definition: `train_qnn`

The training logic is wrapped as an Orchestral AI tool. The agent can specify `epochs` and `learning_rate`, and receives back structured metrics including per-epoch history for learning curve analysis.

In [ ]:
@define_tool()
def train_qnn(epochs: int, learning_rate: float) -> dict:
    """Train a hybrid quantum neural network on MNIST and return performance metrics.

    This tool trains a fixed-architecture hybrid QNN (classical encoder + 4-qubit VQC
    + classical head) on a subset of MNIST (800 train / 200 test images, 10 classes).
    The model uses Adam optimizer and cross-entropy loss.

    After training, it returns per-epoch metrics (train_loss, train_acc, test_acc)
    so you can analyze learning dynamics and decide if the model needs more training.

    Args:
        epochs: Number of training epochs (recommended: 2-10, more epochs = longer training)
        learning_rate: Learning rate for Adam optimizer (recommended: 0.001-0.05)

    Returns:
        Dictionary with:
        - final_train_loss: Training loss after the last epoch
        - final_train_acc: Training accuracy after the last epoch (0-1)
        - final_test_acc: Test accuracy after the last epoch (0-1)
        - history: List of dicts, one per epoch, each with {epoch, train_loss, train_acc, test_acc}
        - training_time_seconds: Total wall-clock training time
        - model_summary: Description of the model architecture
        - recommendation: Whether model needs more training based on the metrics
    """
    torch.manual_seed(0)
    model = HybridQNN()
    optimizer = optim.Adam(model.parameters(), lr=learning_rate)
    criterion = nn.CrossEntropyLoss()

    history = []
    start_time = time.time()

    for epoch in range(1, epochs + 1):
        # Training
        model.train()
        total_loss = 0.0
        correct = 0
        total = 0

        for x_batch, y_batch in TRAIN_LOADER:
            optimizer.zero_grad()
            logits = model(x_batch)
            loss = criterion(logits, y_batch)
            loss.backward()
            optimizer.step()

            total_loss += loss.item() * x_batch.size(0)
            correct += (logits.argmax(dim=1) == y_batch).sum().item()
            total += x_batch.size(0)

        train_loss = total_loss / total
        train_acc = correct / total

        # Evaluation
        model.eval()
        correct = 0
        total = 0
        with torch.no_grad():
            for x_batch, y_batch in TEST_LOADER:
                logits = model(x_batch)
                correct += (logits.argmax(dim=1) == y_batch).sum().item()
                total += x_batch.size(0)

        test_acc = correct / total

        history.append({
            "epoch": epoch,
            "train_loss": round(train_loss, 4),
            "train_acc": round(train_acc, 4),
            "test_acc": round(test_acc, 4),
        })
        print(f"  Epoch {epoch}/{epochs}: loss={train_loss:.4f}, train_acc={train_acc:.4f}, test_acc={test_acc:.4f}")

    elapsed = time.time() - start_time

    final = history[-1]

    # Generate recommendation
    if final["test_acc"] < 0.15:
        recommendation = "Model is performing near random chance (10%). Consider increasing epochs or adjusting learning rate."
    elif final["test_acc"] < 0.30:
        recommendation = "Model shows some learning but accuracy is low. More epochs or a different learning rate may help."
    elif final["train_loss"] > 2.0:
        recommendation = "Training loss is still high. The model would likely benefit from more epochs."
    elif final["train_acc"] - final["test_acc"] > 0.15:
        recommendation = "Signs of overfitting (train_acc >> test_acc). Consider fewer epochs or regularization."
    else:
        recommendation = "Model is training well. Consider more epochs if test accuracy is still improving."

    return {
        "final_train_loss": final["train_loss"],
        "final_train_acc": final["train_acc"],
        "final_test_acc": final["test_acc"],
        "history": history,
        "training_time_seconds": round(elapsed, 1),
        "model_summary": f"HybridQNN: 784->64->{N_QUBITS}(VQC 2L RY+CNOT)->{N_QUBITS}->10",
        "recommendation": recommendation,
    }

print("Tool [train_qnn] defined")

Tool [train_qnn] defined


## Direct Tool Verification (No API Calls)

Test the tool directly before using it with the agent.

In [ ]:
print("Training QNN with epochs=3, lr=0.01...")
result = train_qnn.execute(epochs=3, learning_rate=0.01)
print(f"\nTool returned:")
print(result)

Training QNN with epochs=3, lr=0.01...
  Epoch 1/3: loss=2.2928, train_acc=0.1125, test_acc=0.0950
  Epoch 2/3: loss=2.2162, train_acc=0.2125, test_acc=0.2400
  Epoch 3/3: loss=2.1159, train_acc=0.2800, test_acc=0.2800

Tool returned:
{'final_train_loss': 2.1159, 'final_train_acc': 0.28, 'final_test_acc': 0.28, 'history': [{'epoch': 1, 'train_loss': 2.2928, 'train_acc': 0.1125, 'test_acc': 0.095}, {'epoch': 2, 'train_loss': 2.2162, 'train_acc': 0.2125, 'test_acc': 0.24}, {'epoch': 3, 'train_loss': 2.1159, 'train_acc': 0.28, 'test_acc': 0.28}], 'training_time_seconds': 1.0, 'model_summary': 'HybridQNN: 784->64->4(VQC 2L RY+CNOT)->4->10', 'recommendation': 'Model shows some learning but accuracy is low. More epochs or a different learning rate may help.'}


## Agent Demo — Gemini 2.5 Flash + `train_qnn`

Two queries test the agent's ability to:
1. Call `train_qnn` with specified hyperparameters and analyze the results
2. Call `train_qnn` again with different hyperparameters and compare runs

In [ ]:
agent = Agent(
    llm=Gemini(model="gemini-2.5-flash"),
    tools=[train_qnn],
)

# Query 1: Basic training
query1 = (
    "Train the quantum neural network for 3 epochs with a learning rate of 0.01. "
    "After training, report the final test accuracy, analyze the learning curve "
    "(is the loss decreasing? is accuracy improving?), and recommend whether "
    "more epochs would help."
)

print(f"Query 1: {query1}")
print("-" * 60)
response1 = agent.run(query1, max_iterations=10)
print(f"\nAgent Response:\n{response1.text}")
print(f"\n[Cost so far: ${agent.get_total_cost():.4f}]")

# Query 2: Different hyperparameters
query2 = (
    "Now train the QNN again but with 5 epochs and a learning rate of 0.005. "
    "Compare mentally with the previous run (3 epochs, lr=0.01) and tell me "
    "which configuration seems better and why."
)

print(f"\nQuery 2: {query2}")
print("-" * 60)
response2 = agent.run(query2, max_iterations=10)
print(f"\nAgent Response:\n{response2.text}")
print(f"\n[Cost so far: ${agent.get_total_cost():.4f}]")

print(f"\n{'=' * 60}")
print(f"TOTAL API COST: ${agent.get_total_cost():.4f}")
print(f"{'=' * 60}")

────────────────────────────────────────────────────────────
Query 1: Train the quantum neural network for 3 epochs with a learning rate of 0.01. After training, report the final test accuracy, analyze the learning curve (is the loss decreasing? is accuracy improving?), and recommend whether more epochs would help.
────────────────────────────────────────────────────────────
  Epoch 1/3: loss=2.2928, train_acc=0.1125, test_acc=0.0950
  Epoch 2/3: loss=2.2162, train_acc=0.2125, test_acc=0.2400
  Epoch 3/3: loss=2.1159, train_acc=0.2800, test_acc=0.2800

Agent Response:
The QNN was trained for 3 epochs with a learning rate of 0.01.

**Final Test Accuracy:** 28%

**Learning Curve Analysis:**
*   **Loss:** The training loss decreased steadily from 2.2928 in epoch 1 to 2.1159 in epoch 3, indicating that the model is learning.
*   **Accuracy:** Both training accuracy and test accuracy improved over the epochs. Training accuracy increased from 11.25% to 28%, and test accuracy increased from 9

## Results Summary

### Training Results

| Config | Epochs | LR | Final Test Acc | Final Train Loss | Recommendation |
|--------|--------|-----|---------------|-----------------|----------------|
| Run 1 | 3 | 0.01 | 28.0% | 2.1159 | More epochs or different LR |
| Run 2 | 5 | 0.005 | 39.0% | 1.8540 | Training well, more epochs could help |

### Agent Behavior

| Query | Tool Called | Agent Analyzed Results | Key Insight |
|-------|-----------|----------------------|-------------|
| 1. Train 3ep/lr=0.01 | `train_qnn` | Yes | Loss decreasing, accuracy improving, needs more epochs |
| 2. Train 5ep/lr=0.005 | `train_qnn` | Yes | Compared both runs, correctly identified 5ep/0.005 as better |

**Total API cost:** $0.0009

The agent reliably called `train_qnn` with the specified parameters both times, correctly analyzed the per-epoch metrics, and provided meaningful comparisons between the two training configurations.

---
# Task 3: Hyperparameter Optimization Using Agent

The agent autonomously optimizes the learning rate over 6 trials. The key demonstration is the agent's **decision-making process**: it starts with diverse exploration, identifies the promising range, then narrows down to find the optimal value.

Note: We redefine `train_qnn` below with an extended return format (`converging`, `learning_rate_used`) to help the agent make informed decisions.

## Tool Definition: `train_qnn`

Extended from Task 2 with additional return fields (`converging`, `learning_rate_used`) to help the agent make informed decisions between trials.

In [ ]:
@define_tool()
def train_qnn(epochs: int, learning_rate: float) -> dict:
    """Train a hybrid quantum neural network on MNIST and return performance metrics.

    Trains a fixed-architecture hybrid QNN (classical encoder + 4-qubit VQC + classical head)
    on MNIST (800 train / 200 test). Uses Adam optimizer and cross-entropy loss.

    IMPORTANT: Each call creates a fresh model (random initialization), so results
    reflect the learning rate's effectiveness from scratch. Use this tool to compare
    different learning rates by keeping epochs fixed and varying learning_rate.

    Args:
        epochs: Number of training epochs (recommended: 3-5 for HPO trials)
        learning_rate: Learning rate for Adam optimizer (try values between 0.0001 and 0.1)

    Returns:
        Dictionary with:
        - final_train_loss: Training loss after the last epoch
        - final_train_acc: Training accuracy (0-1)
        - final_test_acc: Test accuracy (0-1) — this is the primary metric to optimize
        - history: Per-epoch list of {epoch, train_loss, train_acc, test_acc}
        - training_time_seconds: Wall-clock time
        - converging: Whether loss is still decreasing in the last epoch
        - learning_rate_used: The learning rate that was used (for your records)
    """
    torch.manual_seed(0)
    model = HybridQNN()
    optimizer = optim.Adam(model.parameters(), lr=learning_rate)
    criterion = nn.CrossEntropyLoss()

    history = []
    start_time = time.time()

    for epoch in range(1, epochs + 1):
        model.train()
        total_loss = 0.0
        correct = 0
        total = 0
        for x_batch, y_batch in TRAIN_LOADER:
            optimizer.zero_grad()
            logits = model(x_batch)
            loss = criterion(logits, y_batch)
            loss.backward()
            optimizer.step()
            total_loss += loss.item() * x_batch.size(0)
            correct += (logits.argmax(dim=1) == y_batch).sum().item()
            total += x_batch.size(0)

        train_loss = total_loss / total
        train_acc = correct / total

        model.eval()
        correct = 0
        total = 0
        with torch.no_grad():
            for x_batch, y_batch in TEST_LOADER:
                logits = model(x_batch)
                correct += (logits.argmax(dim=1) == y_batch).sum().item()
                total += x_batch.size(0)
        test_acc = correct / total

        history.append({
            "epoch": epoch,
            "train_loss": round(train_loss, 4),
            "train_acc": round(train_acc, 4),
            "test_acc": round(test_acc, 4),
        })
        print(f"  [lr={learning_rate}] Epoch {epoch}/{epochs}: loss={train_loss:.4f}, train_acc={train_acc:.4f}, test_acc={test_acc:.4f}")

    elapsed = time.time() - start_time
    final = history[-1]

    # Check if loss is still decreasing
    converging = True
    if len(history) >= 2:
        converging = history[-1]["train_loss"] < history[-2]["train_loss"]

    return {
        "learning_rate_used": learning_rate,
        "final_train_loss": final["train_loss"],
        "final_train_acc": final["train_acc"],
        "final_test_acc": final["test_acc"],
        "history": history,
        "training_time_seconds": round(elapsed, 1),
        "converging": converging,
    }

print("Tool [train_qnn] defined")

Tool [train_qnn] defined


## Agent-Driven Hyperparameter Optimization

The agent is given:
- Budget: **6 trials**
- Fixed: **epochs=5** per trial
- Task: Find the best **learning rate**
- Requirement: Explain reasoning before each trial, analyze after each trial

The agent must demonstrate **autonomous decision-making** based on feedback.

In [ ]:
agent = Agent(
    llm=Gemini(model="gemini-2.5-flash"),
    tools=[train_qnn],
)

hpo_prompt = """You are a hyperparameter optimization agent. Your goal is to find the best
learning rate for a hybrid quantum neural network trained on MNIST.

RULES:
1. You have a budget of exactly 6 trials. Use all 6 trials wisely.
2. For each trial, call the train_qnn tool with epochs=5 and your chosen learning_rate.
3. BEFORE each trial, explain your reasoning: why are you choosing this learning rate?
   What did you learn from previous trials?
4. After each trial, analyze the results: is the model converging? How does test_acc
   compare to previous trials?
5. After all 6 trials, provide a FINAL REPORT with:
   - A table of all trials (trial#, learning_rate, final_test_acc, final_train_loss, converging)
   - The best learning rate and why
   - What you learned about the learning rate landscape for this QNN
   - A brief recommendation for future optimization

Start with a diverse initial exploration, then narrow down based on results.
Suggested starting range: [0.0001, 0.001, 0.005, 0.01, 0.05, 0.1]
But you may deviate from these suggestions based on your analysis."""

response = agent.run(hpo_prompt, max_iterations=45)
print(response.text)
print(f"\nTOTAL API COST: ${agent.get_total_cost():.4f}")

────────────────────────────────────────────────────────────
Agent starting HPO...
────────────────────────────────────────────────────────────
  [lr=0.01] Epoch 1/5: loss=2.2928, train_acc=0.1125, test_acc=0.0950
  [lr=0.01] Epoch 2/5: loss=2.2162, train_acc=0.2125, test_acc=0.2400
  [lr=0.01] Epoch 3/5: loss=2.1159, train_acc=0.2800, test_acc=0.2800
  [lr=0.01] Epoch 4/5: loss=1.9755, train_acc=0.2963, test_acc=0.2700
  [lr=0.01] Epoch 5/5: loss=1.9170, train_acc=0.2612, test_acc=0.2250
  [lr=0.005] Epoch 1/5: loss=2.2777, train_acc=0.1325, test_acc=0.2950
  [lr=0.005] Epoch 2/5: loss=2.2008, train_acc=0.2787, test_acc=0.3100
  [lr=0.005] Epoch 3/5: loss=2.1076, train_acc=0.2700, test_acc=0.3250
  [lr=0.005] Epoch 4/5: loss=1.9722, train_acc=0.3700, test_acc=0.3950
  [lr=0.005] Epoch 5/5: loss=1.8540, train_acc=0.4100, test_acc=0.3900
  [lr=0.001] Epoch 1/5: loss=2.2826, train_acc=0.1812, test_acc=0.1950
  [lr=0.001] Epoch 2/5: loss=2.2207, train_acc=0.2812, test_acc=0.2300
  [lr=0.0

## Results Analysis

### Agent's Trial History

| Trial | Learning Rate | Test Acc | Train Loss | Converging | Agent's Reasoning |
|-------|-------------|----------|-----------|------------|-------------------|
| 1 | 0.01 | 22.5% | 1.917 | Yes | Initial mid-range exploration |
| 2 | 0.005 | 39.0% | 1.854 | Yes | Try lower LR based on trial 1 |
| 3 | 0.001 | 25.0% | 2.127 | Yes | Explore lower bound |
| 4 | **0.0075** | **40.5%** | **1.731** | Yes | Focus between 0.005-0.01 (best range) |
| 5 | 0.009 | 39.0% | 1.663 | Yes | Fine-tune near optimal |
| 6 | 0.006 | 37.0% | 1.825 | Yes | Confirm 0.0075 is peak |

### Agent Decision-Making Quality

The agent demonstrated clear **explore-then-exploit** behavior:

1. **Trials 1-3 (Exploration):** Diverse sampling across [0.001, 0.005, 0.01] to map the landscape
2. **Trials 4-6 (Exploitation):** Narrowed to [0.006, 0.0075, 0.009] around the identified sweet spot
3. **Correct Conclusion:** Agent identified lr=0.0075 as optimal and explained that the landscape has a peak around 0.005-0.009 with dropoff on both sides

### Key: This demonstrates the agent can make **feedback-driven decisions**, which is the core requirement of Task 3.

**Total API cost:** $0.0039

---
## Overall Summary

| Task | Description | Key Result | API Cost |
|------|-------------|-----------|----------|
| 1 | Hello World (4 tools, 5 queries) | All tools reliably called, multi-call success | $0.0032 |
| 2 | QNN Training Tool (2 queries) | Agent analyzed learning curves, compared configs | $0.0009 |
| 3 | Agent-Driven HPO (6 trials) | Found optimal lr=0.0075 (40.5% test acc) | $0.0039 |
| **Total** | | | **$0.0080** |

### Key Takeaways

1. **Orchestral AI + Gemini 2.5 Flash** provides a reliable agent framework for quantum circuit tasks
2. The `@define_tool()` decorator with clear docstrings enables the LLM to correctly select and invoke tools
3. The agent demonstrates genuine **feedback-driven decision-making** in Task 3, using an explore-then-exploit strategy
4. Total cost across all tasks: **$0.008** — highly efficient for the demonstrated capabilities